### PR TITLE
MM-28733 : Admin Advisor v2

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Admin Advisor v2.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-28733

Related Pull Requests
webapp: https://github.com/mattermost/mattermost-webapp/pull/6461
API: https://github.com/mattermost/mattermost-api-reference/pull/574

### CODE REVIEW

**Comment by srkgupta** [comment]
@catalintomai since the messages are still not ready and finalized by @esadur, I will only be able to test the PRs once they are finalised.

**Comment by jasonblais** [comment]
@srkgupta are the messages at https://github.com/mattermost/mattermost-webapp/pull/6461/files#diff-8aed887c744b7eb6f40c0a268ae06e7b?

**Comment by srkgupta** [comment]
@jasonblais Catalin has just added a placeholder for messages. The actual messages are still not finalized by Eric.

File: app/server.go [hunk 1]
```diff
@@ -1187,51 +1187,113 @@ func doSessionCleanup(s *Server) {
...
+
+	if numberOfActiveUsers < model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_25].Limit {
 		return
-	} else if numberOfActiveUsers >= model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_200].Limit && numberOfActiveUsers < model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_400].Limit {
-		warnMetrics = append(warnMetrics, model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_200])
-	} else if numberOfActiveUsers >= model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_400].Limit && numberOfActiveUsers < model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_500].Limit {
-		warnMetrics = append(warnMetrics, model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_400])
-	} else {
-		warnMetrics = append(warnMetrics, model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_500])
+	} else if teamCount >= model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5].Limit && !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5]) {
```
Comment (hahmadia) [inline app/server.go:1246]: Can we break down these long if statements into multilines so they are a little bit more readable?
```suggestion
	} else if teamCount >= model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5].Limit &&
	 !hasBeenAckedOrShown(
	 warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5], 
	 model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_TEAMS_5]
	 ) {
```
Not sure about how golang multiline statements work but currently the way these if statements are formatted make them a little hard to read.

File: app/server.go [hunk 2]
```diff
@@ -1187,51 +1187,113 @@ func doSessionCleanup(s *Server) {
...
+	} else if *a.Config().ServiceSettings.EnableMultifactorAuthentication && !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_MFA], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_MFA]) {
+		warnMetrics = append(warnMetrics, model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_MFA])
+	} else if isDiffEmailAccount && !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_EMAIL_DOMAIN], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_EMAIL_DOMAIN]) {
+		warnMetrics = append(warnMetrics, model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_EMAIL_DOMAIN])
+	}
+
+	// If the system did not cross any of the thresholds for the Contextual Advisories
+	if len(warnMetrics) == 0 {
+		if numberOfActiveUsers >= model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_100].Limit && numberOfActiveUsers < model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_200].Limit && !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_100], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_100]) {
```
Comment (hahmadia) [inline app/server.go:1258]: Same here. Is there anyway these if statements can be formatted to make them a little bit easier to read?

File: app/server.go [hunk 3]
```diff
@@ -1187,51 +1187,113 @@ func doSessionCleanup(s *Server) {
...
+		} else if !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_500], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_500]) {
+			tWarnMetric := model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_ACTIVE_USERS_500]
+			if postsCount > model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_POSTS_500K].Limit && !hasBeenAckedOrShown(warnMetricStatusFromStore[model.SYSTEM_WARN_METRIC_NUMBER_OF_POSTS_500K], model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_POSTS_500K]) {
+				tWarnMetric = model.WarnMetricsTable[model.SYSTEM_WARN_METRIC_NUMBER_OF_POSTS_500K]
+			}
+			warnMetrics = append(warnMetrics, tWarnMetric)
+		}
 	}
 
+	isE0Edition := (model.BuildEnterpriseReady == "true") //license == nil was already validated upstream
```
Comment (hahmadia) [inline app/server.go:1273]:
```suggestion
	isE0Edition := model.BuildEnterpriseReady == "true" //license == nil was already validated upstream
```

File: app/server.go [hunk 4]
```diff
@@ -1242,6 +1304,10 @@ func doCheckNumberOfActiveUsersWarnMetricStatus(a *App) {
 	}
 }
 
+func hasBeenAckedOrShown(metricStatusFromStore string, warnMetric model.WarnMetric) bool {
+	return (metricStatusFromStore == model.WARN_METRIC_STATUS_ACK || (warnMetric.IsBotOnly && metricStatusFromStore == model.WARN_METRIC_STATUS_RUNONCE))
```
Comment (hahmadia) [inline app/server.go:1308]:
```suggestion
	return metricStatusFromStore == model.WARN_METRIC_STATUS_ACK || (warnMetric.IsBotOnly && metricStatusFromStore == model.WARN_METRIC_STATUS_RUNONCE)
```

File: store/sqlstore/user_store.go [hunk 5]
```diff
@@ -1474,6 +1474,14 @@ func (us SqlUserStore) AnalyticsGetInactiveUsersCount() (int64, *model.AppError)
 	return count, nil
 }
 
+func (us SqlUserStore) AnalyticsGetExternalUsers(hostDomain string) (bool, *model.AppError) {
+	count, err := us.GetReplica().SelectInt("SELECT COUNT(Id) FROM Users WHERE Email NOT LIKE :HostDomain", map[string]interface{}{"HostDomain": "%@" + hostDomain})
```
Comment (hahmadia) [inline store/sqlstore/user_store.go:1478]: Is it possible to use squirrel here for our SQL here (https://github.com/Masterminds/squirrel)

**Review by hahmadia — COMMENTED** [review]
Some other general comments (not specific to this PR) but maybe tickets can be made for them?
I noticed that `BuildEnterpriseReady` was a string declared four years ago and I think it's kinda weird how a boolean variable was initialized a string (maybe a reason for this that I'm not sure) but I feel like this should be a boolean and not a string since throughout the code we have many places checking to see if the value is 'true'. 

I think maybe not changing the variable type to boolean but making some sort of function like "BuildEnterpriseReadyAsBoolean" where then we can just do "func model.BuildEnterpriseReadAsBoolean" rather than always needing to do a "model.BuildEnterpriseReady == 'true'". Just my 2 cents and I feel that it could help alleviate the constant "== true" part.

Otherwise all looks good! Thanks

File: app/integration_action.go [hunk 6]
```diff
@@ -443,48 +443,59 @@ func (a *App) doLocalWarnMetricsRequest(rawURL string, upstreamRequest *model.Po
 		HasReactions: true,
 	}
 
-	forceAck := upstreamRequest.Context["force_ack"].(bool)
-
-	if appErr = a.NotifyAndSetWarnMetricAck(warnMetricId, user, forceAck, true); appErr != nil {
-		if forceAck {
-			return appErr
+	isE0Edition := (model.BuildEnterpriseReady == "true") && (a.Srv().License() == nil)
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply app/integration_action.go:446]: would prefer keeping them, find them easier to read and they don't hurt.

File: app/integration_action.go [hunk 7]
```diff
@@ -443,48 +443,59 @@ func (a *App) doLocalWarnMetricsRequest(rawURL string, upstreamRequest *model.Po
...
-				Id:   "emailUs",
-				Name: utils.T("api.server.warn_metric.email_us"),
-				Type: model.POST_ACTION_TYPE_BUTTON,
-				Options: []*model.PostActionOptions{
-					{
-						Text:  "WarnMetricMailtoUrl",
-						Value: mailtoLinkText,
+	} else {
+		botPost.Message = ":white_check_mark: " + utils.T("api.server.warn_metric.bot_response.notification_success.message")
+		forceAck := upstreamRequest.Context["force_ack"].(bool)
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply app/integration_action.go:459]: The way to do type asserts in golang is by placing the type in brackets after the variable name.

File: store/sqlstore/user_store.go (hunk 5 above)
Comment (catalintomai) [reply store/sqlstore/user_store.go:1478]: Possible, but what advantages does squirrel provide in this context. In general I find it easier and more readable to write plain SQL.

File: app/app.go [hunk 8]
```diff
@@ -103,13 +103,50 @@ func (a *App) InitServer() {
...
+		return appErr
+	}
+
+	trialLicenseRequest := &model.TrialLicenseRequest{
+		ServerID:              a.TelemetryId(),
+		Name:                  currentUser.GetDisplayName(model.SHOW_FULLNAME),
+		Email:                 currentUser.Email,
+		SiteName:              *a.Config().TeamSettings.SiteName,
+		SiteURL:               *a.Config().ServiceSettings.SiteURL,
+		Users:                 30,
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply app/app.go:129]: actually I have an open q for the PM if this needs to be dynamic or not.

File: app/server.go (hunk 1 above)
Comment (catalintomai) [reply app/server.go:1246]: in golang the lines can be split only at certain chars (see https://stackoverflow.com/questions/34846848/how-to-break-a-long-line-of-code-in-golang) - which makes it unnatural in this case imo.

File: app/server.go (hunk 2 above)
Comment (catalintomai) [reply app/server.go:1258]: see above.

File: app/server.go (hunk 3 above)
Comment (catalintomai) [reply app/server.go:1273]: I can remove this one.

File: app/server.go (hunk 4 above)
Comment (catalintomai) [reply app/server.go:1308]: I'll rearrange the parenthesis to be consistent with the observation above wrt to readability.

File: store/sqlstore/user_store.go (hunk 5 above)
Comment (hahmadia) [reply store/sqlstore/user_store.go:1478]: Just to keep constant and our move towards using Squirrel 
https://mattermost.atlassian.net/browse/MM-26726

Comment (hahmadia) [reply store/sqlstore/user_store.go:1478]: Also https://mattermost.atlassian.net/browse/MM-24946
```
Mattermost is replacing the SQL queries of the store layer to use the squirrel builder.
```

**Review by hahmadia — APPROVED** [review]
I guess my last comment would be using squirrel instead of plain sql due to our move to using squirrel. Otheriwse, with respect to above comments, everything else looks good.

File: i18n/en.json [hunk 9]
```diff
@@ -8170,4 +8338,4 @@
     "id": "web.incoming_webhook.user.app_error",
     "translation": "Couldn't find the user."
   }
-]
+]
```
Comment (mkraft) [inline i18n/en.json:8341]: There is so much duplication being proposed in this translation file change, but I'm not sure how hard it would be to deal with it and if it's worth it.

File: api4/system.go [hunk 10]
```diff
@@ -722,3 +728,33 @@ func sendWarnMetricAckEmail(c *Context, w http.ResponseWriter, r *http.Request)
...
+
+func requestTrialLicenseAndAckWarnMetric(c *Context, w http.ResponseWriter, r *http.Request) {
+	auditRec := c.MakeAuditRecord("requestTrialLicenseAndAckWarnMetric", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	c.LogAudit("attempt")
+
+	if !c.App.SessionHasPermissionTo(*c.App.Session(), model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply api4/system.go:740]: @esadur - what do you think?

File: api4/system.go [hunk 11]
```diff
@@ -62,6 +62,7 @@ func (api *API) InitSystem() {
 	api.BaseRoutes.ApiRoot.Handle("/restart", api.ApiSessionRequired(restart)).Methods("POST")
 	api.BaseRoutes.ApiRoot.Handle("/warn_metrics/status", api.ApiSessionRequired(getWarnMetricsStatus)).Methods("GET")
 	api.BaseRoutes.ApiRoot.Handle("/warn_metrics/ack/{warn_metric_id:[A-Za-z0-9-_]+}", api.ApiHandler(sendWarnMetricAckEmail)).Methods("POST")
+	api.BaseRoutes.ApiRoot.Handle("/warn_metrics/trial-license-ack/{warn_metric_id:[A-Za-z0-9-_]+}", api.ApiHandler(requestTrialLicenseAndAckWarnMetric)).Methods("POST")
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply api4/system.go:65]: I've seen both approaches being used (e.g. /assign/{user_id:[A-Za-z0-9]+} and others), not sure if there is an officially recommended one.

File: i18n/en.json [hunk 12]
```diff
@@ -2023,45 +2027,209 @@
...
-    "translation": "Acknowledging"
+    "translation": "Contacting Us"
+  },
+  {
+    "id": "api.server.warn_metric.email_domain.contact_us.email_body",
+    "translation": "Mattermost contact us request. I'm interested in learning more about using Guest Accounts.\r\n"
+  },
+  {
+    "id": "api.server.warn_metric.email_domain.notification_body",
+    "translation": "Projects often involve people both inside and outside of your organization. With Guest Accounts, you can bring external partners into your Mattermost system and specify who they can work with and what they can see.\r\n\r\n[Learn more about enabling Guest Accounts](https://www.mattermost.com/docs-guest-accounts/?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=guest-accounts).\r\n\r\nBy clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2047]: re: @esadur and @jasonblais

File: i18n/en.json [hunk 13]
```diff
@@ -2023,45 +2027,209 @@
...
+  },
+  {
+    "id": "api.server.warn_metric.email_domain.notification_title",
+    "translation": "Creating Guest Accounts"
+  },
+  {
+    "id": "api.server.warn_metric.email_domain.start_trial.notification_body",
+    "translation": "Projects often involve people both inside and outside of your organization. With Guest Accounts, you can bring external partners into your Mattermost system and specify who they can work with and what they can see.\r\n\r\n[Learn more about enabling Guest Accounts](https://www.mattermost.com/docs-guest-accounts/?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=guest-accounts)\r\n\r\nBy clicking Start trial, I agree to the [Mattermost Software Evaluation Agreement](https://mattermost.com/software-evaluation-agreement/), [Privacy Policy](https://mattermost.com/privacy-policy/), and receiving product emails."
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2055]: re: @esadur and @jasonblais

File: i18n/en.json [hunk 14]
```diff
@@ -2023,45 +2027,209 @@
...
     "id": "api.server.warn_metric.email_us",
     "translation": "Email us"
   },
+  {
+    "id": "api.server.warn_metric.mfa.contact_us.email_body",
+    "translation": "Mattermost contact us request. I'm interested in learning more about enforcing Multi-Factor Authentication.\r\n"
+  },
+  {
+    "id": "api.server.warn_metric.mfa.notification_body",
+    "translation": "Your Mattermost system has multi-factor authentication enabled, giving users the choice to secure their accounts with additional means of authentication beyond a password. To improve security across the system, you can require all Mattermost accounts to use multi-factor authentication.\r\n\r\n[Learn more about enforcing Multi-Factor Authentication](https://www.mattermost.com/docs-multi-factor-authentication/?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=multi-factor-authentication). \r\n\r\nBy clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2071]: re @esadur and @jasonblais

File: i18n/en.json [hunk 15]
```diff
@@ -2023,45 +2027,209 @@
...
+    "translation": "Your Mattermost system now has 200 users. When you connect Mattermost with your organization's single sign-on provider, users can access Mattermost without having to re-enter their credentials. We recommend you integrate your SAML 2.0 provider with your Mattermost server.[Learn more about integrating with SAML 2.0](https://www.mattermost.com/docs-saml/?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=saml).\r\n\r\nBy clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2111]: re @esadur and @jasonblais

File: i18n/en.json [hunk 16]
```diff
@@ -2023,45 +2027,209 @@
...
+  },
+  {
+    "id": "api.server.warn_metric.number_of_active_users_500.contact_us.email_body",
+    "translation": "Mattermost contact us request. My team now has 500 users, and I'm considering Mattermost Enterprise Edition.\r\n"
   },
   {
     "id": "api.server.warn_metric.number_of_active_users_500.notification_body",
-    "translation": "Mattermost strongly recommends that deployments of over 500 users take advantage of features such as user management, server clustering and performance monitoring. Contact us to learn more and let us know how we can help.\r\n\r\nBy clicking Acknowledge, you'll be sharing your information with Mattermost Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
+    "translation": "Mattermost strongly recommends that deployments of over 500 users take advantage of features such as user management, server clustering and performance monitoring. Contact us to learn more and let us know how we can help.\r\n\r\nBy clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2151]: re @esadur and @jasonblais

File: i18n/en.json [hunk 17]
```diff
@@ -2023,45 +2027,209 @@
...
+    "translation": "Your Enterprise trial is now active. Go to **System Console > User Management > Permissions** to enable Advanced Permissions."
+  },
+  {
+    "id": "api.server.warn_metric.number_of_posts_500k.contact_us.email_body",
+    "translation": "Mattermost contact us request. I'm interested in learning more about improving performance with Elasticsearch.\r\n"
+  },
+  {
+    "id": "api.server.warn_metric.number_of_posts_500k.notification_body",
+    "translation": "Your Mattermost system has a very high number of posts. Mattermost strongly recommends that you implement Elasticsearch in order to optimize search performance and prevent performance degradation and timeouts. Contact us to learn more and let us know how we can help.\r\n\r\n[Learn more about improving performance](https://www.mattermost.com/docs-elasticsearch/?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=elasticsearch)\r\n\r\nBy clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json:2191]: re @esadur and @jasonblais

File: i18n/en.json (hunk 9 above)
Comment (catalintomai) [reply i18n/en.json:8341]: Yeah, there are some reusable strings, however, deciding where to reuse them in code makes things even more hard to follow; having them clearly separated in code and some duplication in the translation file also allows us to easily update the strings as needed.

File: store/sqlstore/user_store.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (catalintomai) [reply store/sqlstore/user_store.go]: Thanks, I think this is addressed now.

File: api4/system.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (esadur) [reply api4/system.go:740]: @mkraft I don't understand the question, what do you mean by not allowing roles with sysconsole access?

File: i18n/en.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json:2047]: @mkraft @catalintomai I'm fine with this suggestion, though I want to keep the "Inc." So it would read "By clicking Contact Us, you'll be sharing your information with Mattermost, Inc." I am changing this in the 'Admin Advisor - Text Copy' doc

File: i18n/en.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json:2055]: @mkraft how about changing it to this: "Projects often involve people both inside and outside of **an** organization." Makes it more universal.

File: i18n/en.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json:2071]: @mkraft good catch, thanks. Fixed it in the text copy Google doc.

File: i18n/en.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json:2111]: @mkraft we know that AD/LDAP and SAML are popular integrations, but we don't have a specific trigger for showing an advisory about them (compared to something like enforcing MFA, where we know folks would have more interest if they've already enabled MFA). We know these features become more useful for admins as the user scale grows and managing accounts becomes more work, and we chose 100 and 200 users with the idea that we can iterate on those thresholds if necessary based on feedback.

File: i18n/en.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json:2151]: @mkraft we use "strongly recommend" for the 500-user count advisory and the Elasticsearch advisory because at that user and message scale, we believe they should be using E10 or E20. Are you suggesting making the phrasing even stronger?

**Review by srkgupta — CHANGES_REQUESTED** [review]
Hi @catalintomai 

Following issues were observed during the testing of this and the dependent PRs

MM-28943 - High user scale advisory is never triggered if an instance has over 500 users and 500k posts

MM-28946 - v2 Admin Advisory should not be triggered if a previous version instance was acknowledged

MM-28947 - "SAML" and "Read-only announcement channels" v2 admin advisory is not triggered

I am still continuing to test the PRs and will add the new issues as a separate comments. Meanwhile I wanted to give you a heads-up so that you can start fixing these issues.

File: i18n/en.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (srkgupta) [reply i18n/en.json:2047]: @catalintomai this needs to be changed here too. Please update it to:

`By clicking Contact Us, you'll be sharing your information with Mattermost, Inc.` instead of `By clicking Contact Us, you'll be sharing your information with Mattermost, Inc., to learn more about upgrading`

**Comment by srkgupta** [comment]
Few more issues:

[MM-28954](https://mattermost.atlassian.net/browse/MM-28954) - Server does not remembers the last advisory notification time and sends a advisory message everytime the server is restarted

[MM-28953](https://mattermost.atlassian.net/browse/MM-28953) - Advisory Banner is not hidden when user contacts by clicking on the button from the Bot Message

[MM-28956](https://mattermost.atlassian.net/browse/MM-28956) - Server continues to show Advisory messages even when user had requested for trial on E0 and request had failed

File: i18n/en.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (srkgupta) [reply i18n/en.json]: @catalintomai this is still not fixed in the latest code. Please fix the same.

**Comment by srkgupta** [comment]
The following issues are fixed, retested and closed:
MM-28946, MM-28954, MM-28956 

MM-28943 is closed as is because the requirements have changed

MM-28947 is deferred for the next release

MM-28953 is partially fixed and has been reopened. This needs to be properly fixed.

File: i18n/en.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (catalintomai) [reply i18n/en.json]: Thanks, should be fixed now.

File: api4/system.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (esadur) [reply api4/system.go:742]: After speaking with @mkraft and @catalintomai, I've added this as a future improvement, but we won't tackle it for 5.28.

File: i18n/en.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json]: We're going to keep the language as-is here. As an org., we strongly recommend that servers with over 500 users upgrade.

File: i18n/en.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (esadur) [reply i18n/en.json]: I rewrote this advisory, and we will change the threshold to be closer to the scale/Elasticsearch guidance in our docs.

**Review by srkgupta — APPROVED** [review]
Tested the updated PR along with it's dependent PRs and all the issues are now fixed and are working fine. Approving the PR.

**Comment by amyblais** [comment]
@mkraft Kind note to review for v5.28.

**Review by mkraft — APPROVED** [review]
